File: TrumpMobile/TrumpMobile/Models/OrderModels.swift
import Firebase
import FirebaseFirestore
import Foundation

struct TrumpOrder: Identifiable, Codable {
  let id: String
  let userId: String
  let planName: String
  let amount: Double
  let orderDate: Date
  let status: OrderStatus
  let billingCompleted: Bool
  let phoneNumber: String?
  let simType: String
  let currentStep: Int?

  enum OrderStatus: String, Codable, CaseIterable {
    case pending = "pending"
    case completed = "completed"
    case cancelled = "cancelled"

    var displayName: String {
      switch self {
      case .pending:
        return "Pending"
      case .completed:
        return "Completed"
      case .cancelled:
        return "Cancelled"
      }
    }
  }
}

struct OrderDetail {
  let orderId: String
  let userId: String
  let status: String
  let orderCompleted: Bool
  let createdAt: Date
  let updatedAt: Date
  let orderCompletionDate: Date?

  // Personal Information
  let firstName: String
  let lastName: String
  let email: String
  let phoneNumber: String

  // Address Information
  let street: String
  let aptNumber: String
  let city: String
  let state: String
  let zip: String
  let country: String

  // Device Information
  let deviceBrand: String
  let deviceModel: String
  let deviceIsCompatible: Bool
  let isForThisDevice: Bool
  let imei: String

  // Service Information
  let numberType: String
  let selectedPhoneNumber: String
  let simType: String
  let portInSkipped: Bool
  let showQRCode: Bool

  // Billing Information
  let billingDetails: String
  let creditCardNumber: String

  init(from data: [String: Any]) {
    self.orderId = data["orderId"] as? String ?? ""
    self.userId = data["userId"] as? String ?? ""
    self.status = data["status"] as? String ?? ""
    self.orderCompleted = data["orderCompleted"] as? Bool ?? false

    // Handle timestamps
    if let createdTimestamp = data["createdAt"] as? Timestamp {
      self.createdAt = createdTimestamp.dateValue()
    } else {
      self.createdAt = Date()
    }

    if let updatedTimestamp = data["updatedAt"] as? Timestamp {
      self.updatedAt = updatedTimestamp.dateValue()
    } else {
      self.updatedAt = Date()
    }

    if let completionTimestamp = data["orderCompletionDate"] as? Timestamp {
      self.orderCompletionDate = completionTimestamp.dateValue()
    } else {
      self.orderCompletionDate = nil
    }

    // Personal Information
    self.firstName = data["firstName"] as? String ?? ""
    self.lastName = data["lastName"] as? String ?? ""
    self.email = data["email"] as? String ?? ""
    self.phoneNumber = data["phoneNumber"] as? String ?? ""

    // Address Information
    self.street = data["street"] as? String ?? ""
    self.aptNumber = data["aptNumber"] as? String ?? ""
    self.city = data["city"] as? String ?? ""
    self.state = data["state"] as? String ?? ""
    self.zip = data["zip"] as? String ?? ""
    self.country = data["country"] as? String ?? ""

    // Device Information
    self.deviceBrand = data["deviceBrand"] as? String ?? ""
    self.deviceModel = data["deviceModel"] as? String ?? ""
    self.deviceIsCompatible = data["deviceIsCompatible"] as? Bool ?? false
    self.isForThisDevice = data["isForThisDevice"] as? Bool ?? false
    self.imei = data["imei"] as? String ?? ""

    // Service Information
    self.numberType = data["numberType"] as? String ?? ""
    self.selectedPhoneNumber = data["selectedPhoneNumber"] as? String ?? ""
    self.simType = data["simType"] as? String ?? ""
    self.portInSkipped = data["portInSkipped"] as? Bool ?? false
    self.showQRCode = data["showQRCode"] as? Bool ?? false

    // Billing Information
    self.billingDetails = data["billingDetails"] as? String ?? ""
    self.creditCardNumber = data["creditCardNumber"] as? String ?? ""
  }
}

File: linkmobile/lib/models/order_models.dart
import 'package:cloud_firestore/cloud_firestore.dart' show Timestamp;

enum OrderStatus {
  pending,
  completed,
  cancelled;

  String get displayName {
    switch (this) {
      case OrderStatus.pending:
        return 'Pending';
      case OrderStatus.completed:
        return 'Completed';
      case OrderStatus.cancelled:
        return 'Cancelled';
    }
  }
}

class Order {
  final String id;
  final String userId;
  final String planName;
  final double amount;
  final DateTime orderDate;
  final OrderStatus status;
  final bool billingCompleted;
  final String? phoneNumber;
  final String simType;
  final int? currentStep;

  Order({
    required this.id,
    required this.userId,
    required this.planName,
    required this.amount,
    required this.orderDate,
    required this.status,
    required this.billingCompleted,
    this.phoneNumber,
    required this.simType,
    this.currentStep,
  });

  factory Order.fromMap(Map<String, dynamic> map, String id) {
    return Order(
      id: id,
      userId: map['userId'] ?? '',
      planName: map['planName'] ?? '',
      amount: (map['amount'] ?? 0).toDouble(),
      orderDate: (map['orderDate'] as Timestamp?)?.toDate() ?? DateTime.now(),
      status: OrderStatus.values.firstWhere(
        (e) => e.name == (map['status'] ?? 'pending'),
        orElse: () => OrderStatus.pending,
      ),
      billingCompleted: map['billingCompleted'] ?? false,
      phoneNumber: map['phoneNumber'],
      simType: map['simType'] ?? '',
      currentStep: map['currentStep'],
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'userId': userId,
      'planName': planName,
      'amount': amount,
      'orderDate': Timestamp.fromDate(orderDate),
      'status': status.name,
      'billingCompleted': billingCompleted,
      'phoneNumber': phoneNumber,
      'simType': simType,
      'currentStep': currentStep,
    };
  }
}

class OrderDetail {
  // Personal Information
  final String? firstName;
  final String? lastName;
  final String? email;
  final String? phoneNumber;

  // Address Information
  final String? street;
  final String? aptNumber;
  final String? city;
  final String? state;
  final String? zip;
  final String? country;

  // Device Information
  final String? deviceBrand;
  final String? deviceModel;
  final String? imei;
  final bool? deviceIsCompatible;

  // Service Information
  final String? numberType;
  final String? selectedPhoneNumber;
  final String? simType;
  final bool? portInSkipped;

  // Billing Information
  final String? creditCardNumber;
  final String? billingDetails;

  OrderDetail({
    this.firstName,
    this.lastName,
    this.email,
    this.phoneNumber,
    this.street,
    this.aptNumber,
    this.city,
    this.state,
    this.zip,
    this.country,
    this.deviceBrand,
    this.deviceModel,
    this.imei,
    this.deviceIsCompatible,
    this.numberType,
    this.selectedPhoneNumber,
    this.simType,
    this.portInSkipped,
    this.creditCardNumber,
    this.billingDetails,
  });

  factory OrderDetail.fromMap(Map<String, dynamic> map) {
    return OrderDetail(
      firstName: map['firstName'],
      lastName: map['lastName'],
      email: map['email'],
      phoneNumber: map['phoneNumber'],
      street: map['street'],
      aptNumber: map['aptNumber'],
      city: map['city'],
      state: map['state'],
      zip: map['zip'],
      country: map['country'],
      deviceBrand: map['deviceBrand'],
      deviceModel: map['deviceModel'],
      imei: map['imei'],
      deviceIsCompatible: map['deviceIsCompatible'],
      numberType: map['numberType'],
      selectedPhoneNumber: map['selectedPhoneNumber'],
      simType: map['simType'],
      portInSkipped: map['portInSkipped'],
      creditCardNumber: map['creditCardNumber'],
      billingDetails: map['billingDetails'],
    );
  }
}

